### ISSUE TITLE
API problem: It's practically impossible to reregister a level-triggered writable event.

### ISSUE THREAD

**Issue by lcy03406** [carllerche/mio#339]
Background: 
In order to prevent the starvation problem there are two solutions:
use edge-triggered and caching readable events,
or use level-triggered and reregister with readable|writable only when the writing would block.

The level-triggered approach requires a reregister operation, but it's hard to be done using the present API.
Consider the following typical situation:
(1) The server register a readable on a connection and run() an event loop and be notified readable of connection. The EventLoop object is borrowed by run() and 'return'ed to the server's handler as a param of ready() callback.
(2) The server read data from the connection, and process the request, dive deeply into the business logic. I mean, it's not practical to add a &mut EventLoop param to EVERY function of the business logic.
(3) After a complicated computation now the business logic decides to send a bunch of data to the client. But when the sending would block, the server should buffer the data and reregister with readable|writable. But now the EventLoop object is sitting on the stack of ready() as a &mut, so there's no way to create another &mut to it.
(4) So the only possible solution is to pass the &mut EventLoop from ready() to everywhere of the business logic, which is not desirable.

**Comment by carllerche** [comment]
You have a few options here.

1) Represent your business logic as a state machine. You push the event into the state machine, the state transitions, then in your handler, you have a match block over the state to determine what to do (whether or not to re-register the socket with the event loop)

2) You use a higher level abstraction that comes w/ the cost of some allocations.

3) You define your business logic on a type `Foo(&'a mut MyHandler, &'a mut EventLoop<..>)` and then you have access to the event loop in `&self`.

I'm going to close this as `won't fix` as I believe that a more convenient higher level API is out of scope for mio and left to other libraries to build on top of mio.

**Comment by lcy03406** [comment]
Sorry it is closed but I want to reply.
1) Not all business can be a state machine.
And a business state should not bother woundblock, this is too metal level.

2) I don't quite catch what abstraction you mean, but I think I'm near it. Now my solution is, Connections don't need to know the loop, when a write call on a connection fails with wouldblock, an rc of the connection will be put into a queue. The queue will be handled in tick callback. But it's wield I'd say. If the loop were for example some int fd like epoll, and not a &mut, the queue is not needed.

3) Such Foo need to be construct in every ready callback, because Foo cannot outlive the &mut EventLoop. And it must be passed over every function that may write data or may call a function that writes data. Renaming "&'a mut EventLoop" to "Foo<'a>" looks better but does not solve anything.

In summary, &mut is too strong. A low level Metal IO library should not constrain the pattern of upper-layer implementations.
